### PR TITLE
Handle bad HTTP requests

### LINKED ISSUE #23034: index_not_found_exception when passing in a long list of index names
I'm on 5.1.2 and I just noticed locally that my tests started to fail due to the length of the index names being passed. **Obviously I should be using aliases for this......**

```
POST /test-events-2016.08.12,test-events-2016.08.13,test-events-2016.08.14,test-events-2016.08.15,test-events-2016.08.16,test-events-2016.08.17,test-events-2016.08.18,test-events-2016.08.19,test-events-2016.08.20,test-events-2016.08.21,test-events-2016.08.22,test-events-2016.08.23,test-events-2016.08.24,test-events-2016.08.25,test-events-2016.08.26,test-events-2016.08.27,test-events-2016.08.28,test-events-2016.08.29,test-events-2016.08.30,test-events-2016.08.31,test-events-2016.09.01,test-events-2016.09.02,test-events-2016.09.03,test-events-2016.09.04,test-events-2016.09.05,test-events-2016.09.06,test-events-2016.09.07,test-events-2016.09.08,test-events-2016.09.09,test-events-2016.09.10,test-events-2016.09.11,test-events-2016.09.12,test-events-2016.09.13,test-events-2016.09.14,test-events-2016.09.15,test-events-2016.09.16,test-events-2016.09.17,test-events-2016.09.18,test-events-2016.09.19,test-events-2016.09.20,test-events-2016.09.21,test-events-2016.09.22,test-events-2016.09.23,test-events-2016.09.24,test-events-2016.09.25,test-events-2016.09.26,test-events-2016.09.27,test-events-2016.09.28,test-events-2016.09.29,test-events-2016.09.30,test-events-2016.10.01,test-events-2016.10.02,test-events-2016.10.03,test-events-2016.10.04,test-events-2016.10.05,test-events-2016.10.06,test-events-2016.10.07,test-events-2016.10.08,test-events-2016.10.09,test-events-2016.10.10,test-events-2016.10.11,test-events-2016.10.12,test-events-2016.10.13,test-events-2016.10.14,test-events-2016.10.15,test-events-2016.10.16,test-events-2016.10.17,test-events-2016.10.18,test-events-2016.10.19,test-events-2016.10.20,test-events-2016.10.21,test-events-2016.10.22,test-events-2016.10.23,test-events-2016.10.24,test-events-2016.10.25,test-events-2016.10.26,test-events-2016.10.27,test-events-2016.10.28,test-events-2016.10.29,test-events-2016.10.30,test-events-2016.10.31,test-events-2016.11.01,test-events-2016.11.02,test-events-2016.11.03,test-events-2016.11.04,test-events-2016.11.05,test-events-2016.11.06,test-events-2016.11.07,test-events-2016.11.08,test-events-2016.11.09,test-events-2016.11.10,test-events-2016.11.11,test-events-2016.11.12,test-events-2016.11.13,test-events-2016.11.14,test-events-2016.11.15,test-events-2016.11.16,test-events-2016.11.17,test-events-2016.11.18,test-events-2016.11.19,test-events-2016.11.20,test-events-2016.11.21,test-events-2016.11.22,test-events-2016.11.23,test-events-2016.11.24,test-events-2016.11.25,test-events-2016.11.26,test-events-2016.11.27,test-events-2016.11.28,test-events-2016.11.29,test-events-2016.11.30,test-events-2016.12.01,test-events-2016.12.02,test-events-2016.12.03,test-events-2016.12.04,test-events-2016.12.05,test-events-2016.12.06,test-events-2016.12.07,test-events-2016.12.08,test-events-2016.12.09,test-events-2016.12.10,test-events-2016.12.11,test-events-2016.12.12,test-events-2016.12.13,test-events-2016.12.14,test-events-2016.12.15,test-events-2016.12.16,test-events-2016.12.17,test-events-2016.12.18,test-events-2016.12.19,test-events-2016.12.20,test-events-2016.12.21,test-events-2016.12.22,test-events-2016.12.23,test-events-2016.12.24,test-events-2016.12.25,test-events-2016.12.26,test-events-2016.12.27,test-events-2016.12.28,test-events-2016.12.29,test-events-2016.12.30,test-events-2016.12.31,test-events-2017.01.01,test-events-2017.01.02,test-events-2017.01.03,test-events-2017.01.04,test-events-2017.01.05,test-events-2017.01.06,test-events-2017.01.07,test-events-2017.01.08,test-events-2017.01.09,test-events-2017.01.10,test-events-2017.01.11,test-events-2017.01.12,test-events-2017.01.13,test-events-2017.01.14,test-events-2017.01.15,test-events-2017.01.16,test-events-2017.01.17,test-events-2017.01.18,test-events-2017.01.19,test-events-2017.01.20,test-events-2017.01.21,test-events-2017.01.22,test-events-2017.01.23,test-events-2017.01.24,test-events-2017.01.25,test-events-2017.01.26,test-events-2017.01.27,test-events-2017.01.28,test-events-2017.01.29,test-events-2017.01.30,test-events-2017.01.31,test-events-2017.02.01,test-events-2017.02.02,test-events-2017.02.03,test-events-2017.02.04,test-events-2017.02.05,test-events-2017.02.06,test-events-2017.02.07/events/_search?ignore_unavailable=true
{
  "size": 11,
  "sort": [
    {
      "date": {
        "order": "desc"
      }
    }
  ],
  "_source": {
    "excludes": [
      "idx",
      "ip",
      "error"
    ]
  },
  "query": {
    "bool": {
      "filter": [
        {
          "bool": {
            "must": [
              {
                "range": {
                  "date": {
                    "gte": "2016-08-12T00:00:00Z",
                    "lte": "2017-02-07T23:53:14.9558531Z"
                  }
                }
              },
              {
                "term": {
                  "is_deleted": {
                    "value": false
                  }
                }
              },
              {
                "term": {
                  "stack_id": {
                    "value": "1ecd0826e447a44e78877ab1"
                  }
                }
              }
            ],
            "must_not": [
              {
                "ids": {
                  "values": [
                    "589a5deaf8a1031b5ce541cc"
                  ]
                }
              }
            ]
          }
        }
      ]
    }
  }
}
````

```
{
  "error": {
    "root_cause": [
      {
        "type": "index_not_found_exception",
        "reason": "no such index",
        "resource.type": "index_or_alias",
        "resource.id": "bad-request",
        "index_uuid": "_na_",
        "index": "bad-request"
      }
    ],
    "type": "index_not_found_exception",
    "reason": "no such index",
    "resource.type": "index_or_alias",
    "resource.id": "bad-request",
    "index_uuid": "_na_",
    "index": "bad-request"
  },
  "status": 404
}
```

### PR DESCRIPTION
**Pull request by jasontedor**
When Netty decodes a bad HTTP request, it marks the decoder result on the HTTP request as a failure, and reroutes the request to GET /bad-request. This either leads to puzzling responses when a bad request is sent to Elasticsearch (if an index named "bad-request" does not exist then it produces an index not found exception and otherwise responds with the index settings for the index named "bad-request"). This commit addresses this by inspecting the decoder result on the HTTP request and dispatching the request to a bad request handler preserving the initial cause of the bad request and providing an error message to the client.

Closes #23034

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/http/HttpServerTransport.java [hunk 1]
```diff
@@ -35,15 +35,31 @@
...
-         * @param threadContext the nodes thread context
+         *
+         * @param request       the request to dispatch
+         * @param channel       the response channel of this request
+         * @param threadContext the thread context
+         */
+        void dispatchRequest(RestRequest request, RestChannel channel, ThreadContext threadContext);
+
+        /**
+         * Dispatches the {@link RestRequest} to the bad request handler.
```
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/http/HttpServerTransport.java]: nit: can you make "bad request" more concrete/descriptive or give a example

**Review by jaymode — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/http/HttpServerTransport.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/http/HttpServerTransport.java]: I pushed c617b7ebbb6cc75f98c8df02e18645f6f673cd35.

**Comment by jasontedor** [comment]
Thanks @jaymode.

**Pull request merged**